sql/src/main/java/io/crate/execution/engine/indexing/ShardDMLExecutor.java

```diff
@@ -155,12 +166,40 @@ private static String getLocalNodeId(ClusterService clusterService) {
         return nodeId;
     }
 
-    private static long processShardResponse(ShardResponse shardResponse) {
+    private static <A> A processResponse(ShardResponse shardResponse, Function<ShardResponse, A> f) {
         Exception failure = shardResponse.failure();
         if (failure != null) {
             Throwables.throwIfUnchecked(failure);
             throw new RuntimeException(failure);
         }
-        return shardResponse.successRowCount();
+        return f.apply(shardResponse);
+    }
+
+    private static Long toRowCount(ShardResponse shardResponse) {
+        return Long.valueOf(processResponse(shardResponse, ShardResponse::successRowCount));
+    }
+
+    private static List<Object[]> toResultRows(ShardResponse shardResponse) {
+        return processResponse(shardResponse, ShardResponse::getResultRows);
     }
+
+    public static final Collector<ShardResponse, long[], Iterable<Row>> ROW_COUNT_COLLECTOR = Collector.of(
```

Comment:
I am a bit sad to reinvent the wheel and not be able to use `Collectors.summingLong()` here, but I am depending in ShardDMLExecutor on the accumulator type.